packages/core/src/client.ts

```diff
@@ -158,22 +164,101 @@ export interface ClientOptions {
 	ws: WebSocketManager;
 }
 
-export function createClient({ rest, ws }: ClientOptions) {
-	const api = new API(rest);
-	const emitter = new AsyncEventEmitter<ManagerShardEventsMap>();
+export class Client extends AsyncEventEmitter<ManagerShardEventsMap> {
+	public readonly rest: REST;
 
-	function wrapIntrinsicProps<T>(obj: T, shardId: number): WithIntrinsicProps<T> {
+	public readonly ws: WebSocketManager;
+
+	public readonly api: API;
+
+	public constructor({ rest, ws }: ClientOptions) {
+		super();
+		this.rest = rest;
+		this.ws = ws;
+		this.api = new API(rest);
+
+		this.ws.on(WebSocketShardEvents.Dispatch, ({ data: dispatch, shardId }) => {
+			// @ts-expect-error event props can't be resolved properly, but they are correct
+			this.emit(dispatch.t, this.wrapIntrinsicProps(dispatch.d, shardId));
+		});
+	}
+
+	/**
+	 * Requests guild members from the gateway.
+	 *
+	 * @see {@link https://discord.com/developers/docs/topics/gateway-events#request-guild-members}
+	 * @param shardId - The id of the shard to request guild members from
+	 * @param options - The options for the request
+	 */
+	public async requestGuildMembers(shardId: number, options: GatewayRequestGuildMembersData) {
+		const promise = new Promise<APIGuildMember[]>((resolve, reject) => {
+			const guildMembers: APIGuildMember[] = [];
+
+			const timeout = setTimeout(() => {
+				reject(new Error('Request timed out'));
+			}, 10_000);
+
+			const handler = ({ shardId: eventShardId, data }: MappedEvents[GatewayDispatchEvents.GuildMembersChunk][0]) => {
+				if (eventShardId !== shardId && data.guild_id !== options.guild_id) {
```

Comment:
```suggestion
				if (eventShardId !== shardId || data.guild_id !== options.guild_id) {
```

Reply:
Also, shouldn't we check for nonce as well? Is the shardId check even relevant, considering guilds have only 1 shard?